src/main.js

```diff
@@ -9,293 +9,316 @@ const ipc = electron.ipcMain;
-    // Dereference the window object, usually you would store windows
-    // in an array if your app supports multi windows, this is the time
-    // when you should delete the corresponding element.
-    mainWindow = null;
-  });
+    var saveWindowState = function(file, window) {
+      var window_state = window.getBounds();
+      window_state.maximized = window.isMaximized();
+      window_state.fullscreen = window.isFullScreen();
+      try {
+        fs.writeFileSync(bounds_info_path, JSON.stringify(window_state));
+      }
+      catch (e) {
+        // [Linux] error happens only when the window state is changed before the config dir is creatied.
+      }
+    };
 
-  // Deny drag&drop navigation in mainWindow.
-  // Drag&drop is allowed in webview of index.html.
-  mainWindow.webContents.on('will-navigate', function(event, url) {
-    var dirname = __dirname;
-    if (process.platform === 'win32') {
-      dirname = '/' + dirname.replace(/\\/g, '/');
-    }
+    mainWindow.on('close', function(event) {
+      if (willAppQuit) { // when [Ctrl|Cmd]+Q
+        saveWindowState(bounds_info_path, mainWindow);
+      }
+      else { // Minimize or hide the window for close button.
+        event.preventDefault();
+        switch (process.platform) {
+          case 'win32':
+          case 'linux':
+            mainWindow.minimize();
+            break;
+          case 'darwin':
+            mainWindow.hide();
+            break;
+          default:
+        }
+      }
+    });
 
-    var index = url.indexOf('file://' + dirname);
-    if (index !== 0) {
-      event.preventDefault();
-    }
+    // App should save bounds when a window is closed.
+    // However, 'close' is not fired in some situations(shutdown, ctrl+c)
+    // because main process is killed in such situations.
+    // 'blur' event was effective in order to avoid this.
+    // Ideally, app should detect that OS is shutting down.
+    mainWindow.on('blur', function() {
+      saveWindowState(bounds_info_path, mainWindow);
+    });
+
+    var app_menu = appMenu.createMenu(mainWindow);
+    Menu.setApplicationMenu(app_menu);
+
+    // Emitted when the window is closed.
+    mainWindow.on('closed', function() {
+      // Dereference the window object, usually you would store windows
+      // in an array if your app supports multi windows, this is the time
+      // when you should delete the corresponding element.
+      mainWindow = null;
+    });
+
+    // Deny drag&drop navigation in mainWindow.
+    // Drag&drop is allowed in webview of index.html.
+    mainWindow.webContents.on('will-navigate', function(event, url) {
+      var dirname = __dirname;
+      if (process.platform === 'win32') {
+        dirname = '/' + dirname.replace(/\\/g, '/');
+      }
+
+      var index = url.indexOf('file://' + dirname);
+      if (index !== 0) {
+        event.preventDefault();
+      }
+    });
   });
-});
+}
+
+// Handle Squirrel on Windows startup events
+switch (process.argv[1]) {
```

Comment:
Please move to the top.

Sure, this part will be called at the first, but events are already registered into app module. So I'm afraid of that some issues happen by depending its timing or other reason. The top of `main.js` looks proper to avoid any other processing.